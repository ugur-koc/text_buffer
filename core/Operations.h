#include <regex>
#include <fstream>
#include <streambuf>
#include <sstream>
#include <stack>

#include "HelperUtils.h"
#include "TextBuffer.h"

class Operation {
public:
    TextBuffer *tb;
    std::regex *arg_regex;
    virtual bool execute(const std::string &args) = 0 ;
    virtual void undo(const std::string &args) = 0;
    virtual void redo(const std::string &args) = 0; // better to have to skip the validity checks for args
};

class Load: public Operation { 
public:
    explicit Load(TextBuffer * _tb) {
        tb = _tb;
        arg_regex = new std::regex(".+");
    }
    bool execute(const std::string &arg) override {
        bool validArgs = std::regex_match(arg, *arg_regex);
        if (!validArgs) {
            HelperUtils::printWarning("Invalid arguments to Load!");
        } else {
            std::ifstream file_stream(arg);
            if (!file_stream) {
                HelperUtils::printWarning("Error in openning file " + arg + "!");
                validArgs = false;
            } else {
                std::stringstream buffer;
                buffer << file_stream.rdbuf();
                tb->setContent(buffer.str());    
            }
        }
        return validArgs;
    }

    void undo(const std::string &arg) override {} // not in use
    void redo(const std::string &arg) override {} // not in use
};

class Save: public Operation { 
public:
    explicit Save(TextBuffer * _tb) {
        tb = _tb;
        arg_regex = new std::regex(".+");
    }

    bool execute(const std::string &arg) override {
        bool validArgs = std::regex_match(arg, *arg_regex);
        if (!validArgs) {
            HelperUtils::printWarning("Invalid arguments to Save!");
        } else {
            std::ofstream file_stream (arg, std::ios::out | std::ios::app | std::ios::binary);
            if (!file_stream.is_open()) {
                HelperUtils::printWarning("Error in openning file " + arg + "!");
                validArgs = false;
            } else {
                file_stream << tb->getContent();
                file_stream.close();
            }
        }
        return validArgs;
    }

    void undo(const std::string &arg) override {} // not in use
    void redo(const std::string &arg) override {} // not in use
};

class Insert: public Operation {
public:
    explicit Insert(TextBuffer * _tb) {
        tb = _tb;
        arg_regex = new std::regex("\\d+\\s.*");
    }

    bool execute(const std::string &arg) override {
        bool validArgs = std::regex_match(arg, *arg_regex);
        if (!validArgs) {
            HelperUtils::printWarning("Invalid arguments to insert!");
        } else {
            int indx = arg.find(" ");
            int pos = stoi(arg.substr(0, indx));
            if (pos > tb->getLength()) {
                validArgs = false;
                HelperUtils::printWarning("Invalid insertion position!");
            } else {
                std::string insertion = arg.substr(indx + 1); // test 'insert 1 '
                tb->insert(pos, insertion); // the actual operation
            }
        }
        return validArgs;
    }

    void undo(const std::string &arg) override {
        // arg retrieved from history, no need for validity check
        int indx = arg.find(" ");
        int pos = stoi(arg.substr(0, indx));
        int insertion_len = arg.length() - indx - 1;
        tb->erase(pos, insertion_len);
    }
    void redo(const std::string &arg) override {
        // arg retrieved from history, no need for validity check
        int indx = arg.find(" ");
        int pos = stoi(arg.substr(0, indx));
        std::string insertion = arg.substr(indx + 1);
        tb->insert(pos, insertion);
    }
};

class Append: public Operation {
public:
    explicit Append(TextBuffer * _tb) {
        tb = _tb;
        arg_regex = new std::regex(".+");
    }

    bool execute(const std::string &arg) override {
        bool validArgs = std::regex_match(arg, *arg_regex);
        if (!validArgs) {
            HelperUtils::printWarning("Nothing to append!");
        } else {
            tb->append(arg);// the actual operation
        }
        return validArgs;
    }

    void undo(const std::string &arg) override {
        // arg retrieved from history, no need for validity check
        int insertion_len = arg.length();
        tb->erase(tb->getLength() - insertion_len, insertion_len);
    }
    void redo(const std::string &arg) override {
        // arg retrieved from history, no need for validity check
        tb->append(arg);
    }
};

class EraseAt: public Operation {
    std::stack<std::string> *erases;
public:
    explicit EraseAt(TextBuffer * _tb) {
        tb = _tb;
        arg_regex = new std::regex("\\d+\\s\\d+");
        erases = new std::stack<std::string>();
    }

    bool execute(const std::string &arg) override {
        bool validArgs = std::regex_match(arg, *arg_regex);
        if (!validArgs) {
            HelperUtils::printWarning("Invalid arguments to EraseAt!");
        } else {
            int indx = arg.find(" ");
            int pos = stoi(arg.substr(0, indx));
            if (pos > tb->getLength()) {
                validArgs = false;
                HelperUtils::printWarning("Invalid EraseAt position!");
            } else {
                int erase_len  = stoi(arg.substr(indx+1));
                std::string removal = tb->getSubString(pos, erase_len);
                erases->push(removal);
                tb->erase(pos, erase_len);// the actual operation
                if ((pos+erase_len)>tb->getLength()) {
                    HelperUtils::printWarning("Erased less then "+std::to_string(erase_len)+" characthers!");
                }
            }
        }
        return validArgs;
    }

    void undo(const std::string &arg) override {
        // arg retrieved from history, no need for validity check
        int indx = arg.find(" ");
        int pos = stoi(arg.substr(0, indx));
        std::string insertion = erases->top();
        erases->pop();
        tb->insert(pos, insertion);
    }
    void redo(const std::string &arg) override {
        // arg retrieved from history, no need for validity check
        int indx = arg.find(" ");
        int pos = stoi(arg.substr(0, indx));
        int erase_len  = stoi(arg.substr(indx+1));
        std::string removal = tb->getSubString(pos, erase_len);
        erases->push(removal);
        tb->erase(pos, erase_len);// the actual operation
    }
};

class Erase: public Operation {
    std::stack<std::string> *erases;
public:
    explicit Erase(TextBuffer * _tb) {
        tb = _tb;
        arg_regex = new std::regex("\\d+");
        erases = new std::stack<std::string>();
    }

    bool execute(const std::string &arg) override {
        bool validArgs = std::regex_match(arg, *arg_regex);
        if (!validArgs) {
            HelperUtils::printWarning("Invalid arguments to Erase!");
        } else {
            int erase_len = stoi(arg);
            int pos = tb->getLength() - erase_len;
            if (pos < 0) {
                HelperUtils::printWarning("Erased less then "+std::to_string(erase_len)+" characthers!");
                pos = 0;
            }
            std::string removal = tb->getSubString(pos, erase_len);
            erases->push(removal);
            tb->erase(pos, erase_len);// the actual operation
        }
        return validArgs;
    }

    void undo(const std::string &arg) override {
        // arg retrieved from history, no need for validity check
        std::string insertion = erases->top();
        erases->pop();
        tb->append(insertion);
    }
    void redo(const std::string &arg) override {
        // arg retrieved from history, no need for validity check
        int erase_len = stoi(arg);
        int pos = tb->getLength() - erase_len;
        if (pos < 0) { pos = 0; }
        std::string removal = tb->getSubString(pos, erase_len);
        erases->push(removal);
        tb->erase(pos, erase_len);// the actual operation
    }
};

class Replace: public Operation {
public:
    explicit Replace(TextBuffer * _tb) {
        tb = _tb;
        arg_regex = new std::regex(".+\\s.*");
    }

    bool execute(const std::string &arg) override {
        bool validArgs = std::regex_match(arg, *arg_regex);
        if (!validArgs) {
            HelperUtils::printWarning("Invalid arguments to Replace!");
        } else {
            int indx = arg.find(" ");
            std::string findStr = arg.substr(0, indx);
            std::string replaceStr = arg.substr(indx + 1);
            tb->replace(findStr,replaceStr);// the actual operation
        }
        return validArgs;
    }

    void undo(const std::string &arg) override {
        // arg retrieved from history, no need for validity check
        int indx = arg.find(" ");
        std::string findStr = arg.substr(0, indx);
        std::string replaceStr = arg.substr(indx + 1);
        tb->replace(replaceStr,findStr);// the actual operation
    }
    void redo(const std::string &arg) override {
        // arg retrieved from history, no need for validity check
        int indx = arg.find(" ");
        std::string findStr = arg.substr(0, indx);
        std::string replaceStr = arg.substr(indx + 1);
        tb->replace(findStr,replaceStr);// the actual operation
    }
};