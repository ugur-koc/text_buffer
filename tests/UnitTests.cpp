/**
    MathWorks Interview Asn 2, UnitTests.cpp
    Purpose: Unit test cases for available operations

    @author Ugur Koc
    @version 0.1 01/25/19 
*/


#include <assert.h>

#include "../core/Operations.h"

void test_load_good(){
	std::string str;
	TextBuffer tb(str);
	Load op(&tb);
	
	// simple load with existing file
	std::string cmd_args = "aaa.txt";
	assert(op.execute(cmd_args));
	assert(tb.getContent() == "I love you");

	op.undo(cmd_args); // does nothing
	assert(tb.getContent() == "I love you");
	op.redo(cmd_args); // does nothing
	assert(tb.getContent() == "I love you");
}

void test_load_bad(){
	std::string str;
	TextBuffer tb(str);
	Load op(&tb);

	// load with non existing file
	assert(!op.execute("bbb.txt"));

	// load with non .txt file
	assert(!op.execute("main.cpp"));

}


void test_save_good(){
	std::string str="I love you";
	TextBuffer tb(str);
	Save op(&tb);
	TextBuffer tb2("");
	Load op2(&tb2);

	// simple save with existing file
	std::string cmd_args = "tests/ccc.txt";
	assert(op.execute(cmd_args));
	assert(op2.execute(cmd_args));
	assert(tb2.getContent() == "I love you");

	op.undo(cmd_args); // does nothing
	assert(tb.getContent() == "I love you");
	op.redo(cmd_args); // does nothing
	assert(tb.getContent() == "I love you");
}

void test_save_bad(){
	std::string str="I love you";
	TextBuffer tb(str);
	Save op(&tb);

	// save to non txt extension file
	assert(!op.execute("main.cpp")); // print out warning

	// save to .txt file
	assert(!op.execute(".txt")); // print out warning
}

void test_insert_good(){
	std::string str = "I love you";
	TextBuffer tb(str);
	Insert op(&tb);
	
	// simple insertion with valid arguments
	std::string cmd_args = "2 do ";
	assert(op.execute(cmd_args));
	assert(tb.getContent() == "I do love you");

	// undo and redo previous insertion
	op.undo(cmd_args);
	assert(tb.getContent() == str);
	op.redo(cmd_args);
	assert(tb.getContent() == "I do love you");
	op.undo(cmd_args); //revert to orig
	
	// insertion to end
	cmd_args = "10  my girl";
	assert(op.execute(cmd_args));
	assert(tb.getContent() == "I love you my girl");
	
	// insertion to the begining
	cmd_args = "0 Why do ";
	assert(op.execute(cmd_args));
	assert(tb.getContent() == "Why do I love you my girl");
}

void test_insert_bad(){
	std::string str = "I love you";
	TextBuffer tb(str);
	Insert op(&tb);
	
	// insertion with negative position
	assert(!op.execute("-2 do ")); // print out warning message
	assert(tb.getContent() == str);

	// insertion with position bigger than buffer lenght
	assert(!op.execute("200 do ")); // print out warning message
	assert(tb.getContent() == str);

	// insertion with non integer position
	assert(!op.execute("pos do ")); // print out warning message
	assert(tb.getContent() == str);

	// insertion with no insertion text
	assert(!op.execute("2")); // print out warning message
	assert(tb.getContent() == str);
}

void test_append_good(){
	std::string str = "I love you";
	TextBuffer tb(str);
	Append op(&tb);
	
	// simple append with valid arguments
	std::string cmd_args = " my girl";
	assert(op.execute(cmd_args));
	assert(tb.getContent() == "I love you my girl");

	// undo and redo previous append
	op.undo(cmd_args);
	assert(tb.getContent() == str);
	op.redo(cmd_args);
	assert(tb.getContent() == "I love you my girl");
	op.undo(cmd_args); //revert to orig

	// append space
	cmd_args = " ";
	assert(op.execute(cmd_args)); // print out warning message
	assert(tb.getContent() == "I love you ");
	op.undo(cmd_args); //revert to orig
}

void test_append_bad(){
	std::string str = "I love you";
	TextBuffer tb(str);
	Append op(&tb);

	// append empty text
	std::string cmd_args = "";
	assert(!op.execute(cmd_args)); // print out warning message
	assert(tb.getContent() == "I love you");

}

void test_eraseAt_good(){
	std::string str = "I love you";
	TextBuffer tb(str);
	EraseAt op(&tb);
	
	// simple eraseAt with valid arguments
	std::string cmd_args = "2 5";
	assert(op.execute(cmd_args));
	assert(tb.getContent() == "I you");

	// undo and redo previous eraseAt
	op.undo(cmd_args);
	assert(tb.getContent() == str);
	op.redo(cmd_args);
	assert(tb.getContent() == "I you");
	op.undo(cmd_args); //revert to orig
	
	// eraseAt end
	cmd_args = "6 4";
	assert(op.execute(cmd_args));
	assert(tb.getContent() == "I love");
	op.undo(cmd_args); //revert to orig

	// eraseAt the begining
	cmd_args = "0 2";
	assert(op.execute(cmd_args));
	assert(tb.getContent() == "love you");
}

void test_eraseAt_bad(){
	std::string str = "I love you";
	TextBuffer tb(str);
	EraseAt op(&tb);
	
	// eraseAt negative position
	assert(!op.execute("-2 3")); // print out warning message
	assert(tb.getContent() == str);

	// eraseAt with position bigger than buffer lenght
	assert(!op.execute("200 3")); // print out warning message
	assert(tb.getContent() == str);

	// eraseAt with non integer position and length
	assert(!op.execute("pos do ")); // print out warning message
	assert(tb.getContent() == str);

	// eraseAt with no length
	assert(!op.execute("2")); // print out warning message
	assert(tb.getContent() == str);

	// eraseAt end shorther lenght
	assert(op.execute("6 5")); // print out warning message
	assert(tb.getContent() == "I love");
}

void test_erase_good(){
	std::string str = "I love you";
	TextBuffer tb(str);
	Erase op(&tb);
	
	// simple erase with valid arguments
	std::string cmd_args = "4";
	assert(op.execute(cmd_args));
	assert(tb.getContent() == "I love");

	// undo and redo previous eraseAt
	op.undo(cmd_args);
	assert(tb.getContent() == str);
	op.redo(cmd_args);
	assert(tb.getContent() == "I love");
	op.undo(cmd_args); //revert to orig
}

void test_erase_bad(){
	std::string str = "I love you";
	TextBuffer tb(str);
	Erase op(&tb);
	
	// erase negative position
	assert(!op.execute("-2")); // print out warning message
	assert(tb.getContent() == str);

	// erase with non integer position and length
	assert(!op.execute("pos")); // print out warning message
	assert(tb.getContent() == str);

	// erase with no length
	assert(!op.execute("")); // print out warning message
	assert(tb.getContent() == str);

	// erase with position bigger than buffer length
	assert(op.execute("200")); // print out warning message
	assert(tb.getContent() == "");
}

void test_replace_good(){
	std::string str = "I love you";
	TextBuffer tb(str);
	Replace op(&tb);
	
	// simple replace with valid arguments
	std::string cmd_args = "love like";
	assert(op.execute(cmd_args));
	assert(tb.getContent() == "I like you");

	// undo and redo previous replace
	op.undo(cmd_args);
	assert(tb.getContent() == str);
	op.redo(cmd_args);
	assert(tb.getContent() == "I like you");
	op.undo(cmd_args); //revert to orig

	// replace non-occurring substr
	cmd_args = "her you";
	assert(op.execute(cmd_args));
	assert(tb.getContent() == "I love you");
	
	// replace multi-occurring substr
	cmd_args = "o a";
	assert(op.execute(cmd_args));
	assert(tb.getContent() == "I lave yau");
	op.undo(cmd_args); //revert to orig

	// replace with space
	cmd_args = "love  ";
	assert(op.execute(cmd_args));
	assert(tb.getContent() == "I   you");
	op.undo(cmd_args); //revert to orig
	assert(tb.getContent() == "I love you");

	// replace with nothing
	cmd_args = "love ";
	assert(op.execute(cmd_args));
	assert(tb.getContent() == "I  you");
	op.undo(cmd_args); //revert to orig
	assert(tb.getContent() == "I love you");
}

void test_replace_bad(){
	std::string str = "I love you";
	TextBuffer tb(str);
	Replace op(&tb);
	
	// replace no argument
	assert(!op.execute("")); // print out warning message
	assert(tb.getContent() == str);

	// replace single argument
	assert(!op.execute("love")); // print out warning message
	assert(tb.getContent() == str);

	// replace space
	assert(!op.execute("  love")); // print out warning message
	assert(tb.getContent() == str);
}

int main() {
	test_load_good();
	test_load_bad();
	test_save_good();
	test_save_bad();
	test_insert_good();
	test_insert_bad();
	test_append_good();
	test_append_bad();
	test_eraseAt_good();
	test_eraseAt_bad();
	test_erase_good();
	test_erase_bad();
	test_replace_good();
	test_replace_bad();
	return 0;
}
