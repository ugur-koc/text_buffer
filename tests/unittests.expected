Error in openning file bbb.txt!

Invalid arguments to Load!

Invalid arguments to Save!

Invalid arguments to Save!

Invalid arguments to insert!

Invalid insertion position!

Invalid arguments to insert!

Invalid arguments to insert!

Nothing to append!

Invalid arguments to EraseAt!

Invalid EraseAt position!

Invalid arguments to EraseAt!

Invalid arguments to EraseAt!

Erased less than 5 characters!

Invalid arguments to Erase!

Invalid arguments to Erase!

Invalid arguments to Erase!

Erased less than 200 characters!

Invalid arguments to Replace!

Invalid arguments to Replace!

Invalid arguments to Replace!

